zookeeper-server/src/main/java/org/apache/zookeeper/server/PrepRequestProcessor.java

```diff
@@ -1003,6 +1003,45 @@ public static List<ACL> fixupACL(String path, List<Id> authInfo, List<ACL> acls)
             throw new KeeperException.InvalidACLException(path);
         }
         List<ACL> rv = new ArrayList<>();
+
+        // Overwrite the acl list for users (except super user) when znode group acl feature is on;
+        // Set znode ACL to the corresponding znode group acl designated ACL instead.
+        // Examples that will be processed are:
+        // Single domain user / cross domain components -> set (x509 : domainName) as znode ACL
+        // Users whose extracted clientId is not found in the ClientURIDomainMapping
+        //      -> set (x509: clientURI) as znode ACL
+        // Examples that will not be handled by the process are:
+        //      x509 super user, plaintext port clients, any user when connection filtering feature is on
+        //      -> will go through original zk fixupACL logic
+        boolean isProcessed = false;
+        if (X509AuthenticationConfig.getInstance().isX509ClientIdAsAclEnabled()
+            && X509AuthenticationConfig.getInstance().isX509ZnodeGroupAclEnabled()
+            && !X509AuthenticationConfig.getInstance().isZnodeGroupAclDedicatedServerEnabled()) {
+            for (Id id : authInfo) {
+                boolean isX509SchemeId = id.getScheme().equals(X509AuthenticationUtil.X509_SCHEME);
+                boolean isX509CrossDomainComponent =
+                    id.getScheme().equals(X509AuthenticationUtil.SUPERUSER_AUTH_SCHEME) && !(id
+                        .getId()).equals(
+                        X509AuthenticationConfig.getInstance().getZnodeGroupAclSuperUserId());
```

Comment:
Why do we need to wrap `id
                        .getId()).equals(
                        X509AuthenticationConfig.getInstance().getZnodeGroupAclSuperUserId()` with parentheses?